### ISSUE TITLE
TypingError : Failed in nopython mode pipeline (step: nopython frontend)

### ISSUE THREAD

**Issue by Rhuax** [MaartenGr/BERTopic#50]
I'm trying to load a trained BERTopic model from disk by using `BERTopic.load`, but I'm getting this error:

```
TypingError                               Traceback (most recent call last)
<ipython-input-9-2081de8232b3> in <module>
      1 import joblib
      2 with open('bertopic_model', 'rb') as file:
----> 3     model=joblib.load(file)

/mnt/d/maserati/git/ticket_analysis/env/lib/python3.8/site-packages/joblib/numpy_pickle.py in load(filename, mmap_mode)
    573         filename = getattr(fobj, 'name', '')
    574         with _read_fileobject(fobj, filename, mmap_mode) as fobj:
--> 575             obj = _unpickle(fobj)
    576     else:
    577         with open(filename, 'rb') as f:

/mnt/d/maserati/git/ticket_analysis/env/lib/python3.8/site-packages/joblib/numpy_pickle.py in _unpickle(fobj, filename, mmap_mode)
    502     obj = None
    503     try:
--> 504         obj = unpickler.load()
    505         if unpickler.compat_mode:
    506             warnings.warn("The file '%s' has been generated with a "

/usr/lib/python3.8/pickle.py in load(self)
   1208                     raise EOFError
   1209                 assert isinstance(key, bytes_types)
-> 1210                 dispatch[key[0]](self)
   1211         except _Stop as stopinst:
   1212             return stopinst.value

/mnt/d/maserati/git/ticket_analysis/env/lib/python3.8/site-packages/joblib/numpy_pickle.py in load_build(self)
    327         NDArrayWrapper is used for backward compatibility with joblib <= 0.9.
    328         """
--> 329         Unpickler.load_build(self)
    330 
    331         # For backward compatibility, we support NDArrayWrapper objects.

/usr/lib/python3.8/pickle.py in load_build(self)
   1701         setstate = getattr(inst, "__setstate__", None)
   1702         if setstate is not None:
-> 1703             setstate(state)
   1704             return
   1705         slotstate = None

/mnt/d/maserati/git/ticket_analysis/env/lib/python3.8/site-packages/pynndescent/pynndescent_.py in __setstate__(self, d)
   1026     def __setstate__(self, d):
   1027         self.__dict__ = d
-> 1028         self._rp_forest = tuple([renumbaify_tree(tree) for tree in d["_rp_forest"]])
   1029 
   1030     def _init_search_graph(self):

/mnt/d/maserati/git/ticket_analysis/env/lib/python3.8/site-packages/pynndescent/pynndescent_.py in <listcomp>(.0)
   1026     def __setstate__(self, d):
   1027         self.__dict__ = d
-> 1028         self._rp_forest = tuple([renumbaify_tree(tree) for tree in d["_rp_forest"]])
   1029 
   1030     def _init_search_graph(self):

/mnt/d/maserati/git/ticket_analysis/env/lib/python3.8/site-packages/pynndescent/rp_trees.py in renumbaify_tree(tree)
   1176     point_indices = numba.typed.List.empty_list(point_indices_type)
   1177 
-> 1178     hyperplanes.extend(tree.hyperplanes)
   1179     offsets.extend(tree.offsets)
   1180     children.extend(tree.children)

/mnt/d/maserati/git/ticket_analysis/env/lib/python3.8/site-packages/numba/typed/typedlist.py in extend(self, iterable)
    364             # can not be sliced.
    365             self._initialise_list(iterable[0])
--> 366         return _extend(self, iterable)
    367 
    368     def remove(self, item):

/mnt/d/maserati/git/ticket_analysis/env/lib/python3.8/site-packages/numba/core/dispatcher.py in _compile_for_args(self, *args, **kws)
    413                 e.patch_message(msg)
    414 
--> 415             error_rewrite(e, 'typing')
    416         except errors.UnsupportedError as e:
    417             # Something unsupported is present in the user code, add help info

/mnt/d/maserati/git/ticket_analysis/env/lib/python3.8/site-packages/numba/core/dispatcher.py in error_rewrite(e, issue_type)
    356                 raise e
    357             else:
--> 358                 reraise(type(e), e, None)
    359 
    360         argtypes = []

/mnt/d/maserati/git/ticket_analysis/env/lib/python3.8/site-packages/numba/core/utils.py in reraise(tp, value, tb)
     78         value = tp()
     79     if value.__traceback__ is not tb:
---> 80         raise value.with_traceback(tb)
     81     raise value
     82 

TypingError: Failed in nopython mode pipeline (step: nopython frontend)
- Resolution failure for literal arguments:
No implementation of function Function(<function impl_extend at 0x7f2a3dc6f4c0>) found for signature:

 >>> impl_extend(ListType[array(float64, 2d, C)], reflected list(array(float32, 1d, C))<iv=None>)

There are 2 candidate implementations:
  - Of which 2 did not match due to:
  Overload in function 'impl_extend': File: numba/typed/listobject.py: Line 1027.
    With argument(s): '(ListType[array(float64, 2d, C)], reflected list(array(float32, 1d, C))<iv=None>)':
   Rejected as the implementation raised a specific error:
     TypingError: Failed in nopython mode pipeline (step: nopython frontend)
   - Resolution failure for literal arguments:
   No implementation of function Function(<function impl_append at 0x7f2a3dcf4c10>) found for signature:

    >>> impl_append(ListType[array(float64, 2d, C)], array(float32, 1d, C))

   There are 2 candidate implementations:
         - Of which 2 did not match due to:
         Overload in function 'impl_append': File: numba/typed/listobject.py: Line 589.
           With argument(s): '(ListType[array(float64, 2d, C)], array(float32, 1d, C))':
          Rejected as the implementation raised a specific error:
            LoweringError: Failed in nopython mode pipeline (step: nopython mode backend)
          
          
          File "../env/lib/python3.8/site-packages/numba/typed/listobject.py", line 597:
              def impl(l, item):
                  casteditem = _cast(item, itemty)
                  ^
          
          During: lowering "$8call_function.3 = call $2load_global.0(item, $6load_deref.2, func=$2load_global.0, args=[Var(item, listobject.py:597), Var($6load_deref.2, listobject.py:597)], kws=(), vararg=None)" at /mnt/d/maserati/git/ticket_analysis/env/lib/python3.8/site-packages/numba/typed/listobject.py (597)
     raised from /mnt/d/maserati/git/ticket_analysis/env/lib/python3.8/site-packages/numba/core/utils.py:81
   
   - Resolution failure for non-literal arguments:
   None
   
   During: resolving callee type: BoundFunction((<class 'numba.core.types.containers.ListType'>, 'append') for ListType[array(float64, 2d, C)])
   During: typing of call at /mnt/d/maserati/git/ticket_analysis/env/lib/python3.8/site-packages/numba/typed/listobject.py (1051)
   
   
   File "../env/lib/python3.8/site-packages/numba/typed/listobject.py", line 1051:
               def impl(l, iterable):
                   <source elided>
                   for i in iterable:
                       l.append(i)
                       ^

  raised from /mnt/d/maserati/git/ticket_analysis/env/lib/python3.8/site-packages/numba/core/typeinfer.py:1071

- Resolution failure for non-literal arguments:
None

During: resolving callee type: BoundFunction((<class 'numba.core.types.containers.ListType'>, 'extend') for ListType[array(float64, 2d, C)])
During: typing of call at /mnt/d/maserati/git/ticket_analysis/env/lib/python3.8/site-packages/numba/typed/typedlist.py (101)


File "../env/lib/python3.8/site-packages/numba/typed/typedlist.py", line 101:
def _extend(l, iterable):
    return l.extend(iterable)
    ^

```
I tried to upgrade to joblib 1.0.0 but I'm still getting the same error. Did someone receive the same error in the past?
Why not use pickle/dill instead of joblib==0.17.0 ?

**Comment by MaartenGr** [comment]
Although I am not familiar with this specific error message, there are several reasons why loading a previously trained model would not work. First, make sure BERTopic is the same version at the moment of saving the file and loading it later. The API of BERTopic changes significantly over time and might be different between instances. Second, joblib, and any other dependencies, typically also needs to be the same version as there might be large differences in their API structure. Joblib, for example, cannot currently be set at 1.0.0 since the pipy version of HDBSCAN has issues with that version. Therefore, it is important that joblib==0.17.0! 

With respect to pickle vs. joblib, they both use the same protocol and the differences are not that big. So I opted for joblib since it was easier to use for this specific use-case (i.e., BERTopic).

**Comment by Rhuax** [comment]
Thanks for your fast reply! Yes, indeed I tried first joblib==0.17.0 and then joblib==1.0.0. I also found joblib relies on pickle, infact I tried to pickle/unpickle the model, but I'm still having the same issue. 
Speaking about the BERTopic/UMAP library versions, I literally trained the model yesterday, so I didn't change libraries' versions.
I found it is a problem related with numba, I tried to upgrade numba but it still gives me the same error.
I'm trying to found solutions on the internet, so if I find it I'll report it here.
